### ISSUE TITLE
Type error when importing server build from `server.ts`

### ISSUE THREAD

**Issue by ZeldOcarina** [remix-run/remix#8343]
### Reproduction

https://stackblitz.com/edit/remix-run-remix-2pvj8y?file=server.ts

### System Info

```shell
System:
    OS: macOS 14.2
    CPU: (8) arm64 Apple M1
    Memory: 71.86 MB / 8.00 GB
    Shell: 5.9 - /bin/zsh
  Binaries:
    Node: 20.10.0 - ~/.nvm/versions/node/v20.10.0/bin/node
    Yarn: 1.22.19 - ~/.yarn/bin/yarn
    npm: 10.2.3 - ~/.nvm/versions/node/v20.10.0/bin/npm
    bun: 1.0.3 - ~/.bun/bin/bun
  Browsers:
    Chrome: 120.0.6099.129
    Safari: 17.2
  npmPackages:
    @remix-run/dev: ^2.4.0 => 2.4.0 
    @remix-run/eslint-config: ^2.4.0 => 2.4.0 
    @remix-run/express: ^2.4.0 => 2.4.0 
    @remix-run/node: ^2.4.0 => 2.4.0 
    @remix-run/react: ^2.4.0 => 2.4.0 
    vite: ^5.0.10 => 5.0.10
```

### Used Package Manager

npm

### Expected Behavior

Clean TS slate

### Actual Behavior

When I add to custom TS server setup:

```js
app.all(
  "*",
  createRequestHandler({
    build: vite
      ? () => vite.ssrLoadModule(unstable_viteServerBuildModuleId)
      : await import("./build/server/index.js"),
  }),
);
```
```
I get: Type '(() => Promise<Record<string, any>>) | typeof import("/Users/mattiarasulo/Desktop/web-development-desktop.nosync/call-center-directory/build/server/index")' is not assignable to type 'ServerBuild | (() => Promise<ServerBuild>)'.
  Type '() => Promise<Record<string, any>>' is not assignable to type 'ServerBuild | (() => Promise<ServerBuild>)'.
    Type '() => Promise<Record<string, any>>' is not assignable to type '() => Promise<ServerBuild>'.
      Type 'Promise<Record<string, any>>' is not assignable to type 'Promise<ServerBuild>'.
        Type 'Record<string, any>' is missing the following properties from type 'ServerBuild': mode, entry, routes, assets, and 3 more.ts(2322)
server.d.ts(18, 5): The expected type comes from property 'build' which is declared here on type '{ build: ServerBuild | (() => Promise<ServerBuild>); getLoadContext?: GetLoadContextFunction | undefined; mode?: string | undefined; }'
```

**Comment by ZeldOcarina** [comment]
Hey @pcattori I think this has been introduced in the latest 2.4.0 release of `@remix-run/dev`, I'll just add a `@ts-expect-error` for now. Thanks for your help!

**Comment by markdalgleish** [comment]
I'm not sure there's an alternative here. In a fresh project when running the dev server before running the build, `"./build/server/index.js`" doesn't even exist yet, so even if Vite was configured to output types, it would still be a type error. To me it makes sense to use `@ts-expect-error` as you said. Given this is pretty much a one-time setup, I don't think the types are as critical here.

**Comment by ZeldOcarina** [comment]
Yeah, is there any specific reason for not using @ts-expect-error directly instead of needing two comments?

**Comment by pcattori** [comment]
Changed the title and removed the Vite tag since this has always been the case with Remix server builds, even with the classic compiler. Definitely something I'm keen to improve, but just not directly related to Vite.

**Comment by ZeldOcarina** [comment]
Thank you for checking this out @pcattori you are the man, I really appreciate you! If there's anything I can do to contribute, I'd be keen to!

**Comment by TheAifam5** [comment]
Hey, I have the same issue. My use-case is to lint and typechecking with `tsc` and linting with `eslint` before building all assets for the Cloudflare.

Is there any timeline for that? I have tried to use `@ts-expect-error` but that only works when a `build` directory does not exist, otherwise does throw error that is unused - migrated over to `@ts-ignore` for lines where was needed and that solved it.

**Comment by brophdawg11** [comment]
Thank you for opening this issue, and our apologies we haven't gotten around to it yet!

With the release of [React Router v7](https://remix.run/blog/react-router-v7) we are sun-setting continued development/maintenance on Remix v2. If you have not already [upgraded to React Router v7](https://reactrouter.com/upgrading/remix), we recommend you do so. We've tried to make the upgrade process as smooth as possible with our [Future Flags](https://remix.run/docs/en/main/start/future-flags). We are now in the process of cleaning up outdated issues and pull requests to improve the overall hygiene of our repositories.

We plan to continue to address 2 types of issues in Remix v2:

- Bugs that pose security concerns
- Bugs that prevent upgrading to React Router v7

If you believe this issue meets one of those criteria, please respond or [create a new issue](https://github.com/remix-run/remix/issues/new?template=bug_report.yml).

For all other issues, ongoing maintenance will be happening in React Router v7, so:

- If this is a bug, please reopen this issue [in that repo](https://github.com/remix-run/react-router/issues/new?template=bug_report.yml) with a new [minimal reproduction](https://reactrouter.com/new) against v7
- If this is a feature request, please open a new [Proposal Discussion](https://github.com/remix-run/react-router/discussions/new?category=proposals) in React Router, and if it gets enough community support it can be considered for implementation

If you have any questions you can always reach out on [Discord](https://rmx.as/discord). Thanks again for providing feedback and helping us make our framework even better!
